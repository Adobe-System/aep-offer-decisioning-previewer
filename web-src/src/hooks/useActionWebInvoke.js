import { useState, useEffect } from "react";
import actionWebInvoke from "../utils";

export const useActionWebInvoke = ({
  actionName,
  headers = {},
  params = {},
}) => {
  const [data, setData] = useState(null);
  const [isLoading, setIsLoading] = useState(true);
  const [error, setError] = useState(null);

  useEffect(() => {
    // invoke Action
    actionWebInvoke(actionName, headers, params)
      .then((response) => {
        setData(response);
        setIsLoading(false);
      })
      .catch((e) => {
        setData(null);
        setIsLoading(false);
        setError(e);
      });
  }, []);
  return { data, isLoading, error };
};
