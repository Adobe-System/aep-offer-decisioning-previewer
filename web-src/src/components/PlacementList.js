import React from "react";
import PropTypes from "prop-types";
import { Picker, ProgressCircle, Item, Text } from "@adobe/react-spectrum";
import { useActionWebInvoke } from "../hooks/useActionWebInvoke";

const PlacementList = (props) => {
  const enabledPlacements = props.enabledPlacements || [];
  let headers = {};
  if (props.ims.token && !headers.authorization) {
    headers.authorization = `Bearer ${props.ims.token}`;
  }
  if (props.ims.org && !headers["x-gw-ims-org-id"]) {
    headers["x-gw-ims-org-id"] = props.ims.org;
  }

  const placements = useActionWebInvoke({
    actionName: "get-placements",
    headers: headers,
    params: {
      containerID: props.containerID,
    },
  });

  let picker = (
    <ProgressCircle
      id="placement-list-progress-circle"
      aria-label="Getting Offer Placement"
      isIndeterminate
      isHidden={!placements.isLoading}
      marginStart="size-100"
    />
  );

  if (placements.error) {
    picker = <Text>{placements.error.message}</Text>;
  }

  if (!placements.data && !placements.error && !placements.isLoading) {
    picker = <Text>You have no placements !</Text>;
  }
  if (placements.data) {
    const disabledKeys = placements.data._embedded.results
      .filter(
        (placement) => !enabledPlacements.includes(placement._instance["@id"])
      )
      .map((placement) => placement._instance["@id"]);

    picker = (
      <Picker
        id="placement-list-picker"
        width="100%"
        maxWidth="100%"
        label="Select Offer Placement"
        labelPosition="side"
        labelAlign="end"
        isRequired={true}
        placeholder="select a placement"
        aria-label="select a placement"
        items={placements.data._embedded.results.map((placement) => ({
          name: placement._instance["xdm:name"],
          id: placement._instance["@id"],
        }))}
        itemKey="id"
        isDisabled={props.isDisabled}
        selectedKey={props.defaultSelection}
        disabledKeys={disabledKeys}
        onSelectionChange={props.onSelectionChange}
      >
        {(item) => <Item key={item.id}>{item.name}</Item>}
      </Picker>
    );
  }

  return picker;
};

PlacementList.propTypes = {
  runtime: PropTypes.any,
  containerID: PropTypes.string,
  ims: PropTypes.any,
};

export default PlacementList;
