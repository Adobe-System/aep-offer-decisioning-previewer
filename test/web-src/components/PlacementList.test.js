import React from "react";
import { Provider, defaultTheme } from "@adobe/react-spectrum";
import { render, screen, cleanup } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom/extend-expect";
jest.mock("../../../web-src/src/hooks/useActionWebInvoke.js");
import { useActionWebInvoke } from "../../../web-src/src/hooks/useActionWebInvoke";
import mock from "../../actions/mock";

import PlacementList from "../../../web-src/src/components/PlacementList";

afterEach(cleanup);
beforeEach(() => {
  jest.clearAllMocks();
});

const onSelectionChange = jest.fn(0);
const fakeProps = {
  ims: { token: "fake-token", org: "fake-org" },
  containerID: "fake-containerID",
  onSelectionChange: onSelectionChange,
};

describe("<PlacementList> calls custom hook", () => {
  const mockResponse = {
    data: null,
    isLoading: true,
    error: null,
  };
  it("call custom hooks with global parameters correctly", () => {
    const fakeProps = {
      ims: { token: "fake-token", org: "fake-org" },
      containerID: "fake-containerID",
      onSelectionChange: onSelectionChange,
    };

    useActionWebInvoke.mockReturnValue(mockResponse);
    const { asFragment } = render(<PlacementList {...fakeProps} />);
    expect(asFragment()).toMatchSnapshot();
  });
  it("call custom hooks with global parameters correctly", () => {
    const fakeProps = {
      ims: { authorization: "fake-token", "x-gw-ims-org-id": "fake-org" },
      onSelectionChange: onSelectionChange,
    };

    useActionWebInvoke.mockReturnValue(mockResponse);
    const { asFragment } = render(<PlacementList {...fakeProps} />);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe("<PlacementList> on load", () => {
  const mockResponse = {
    data: null,
    isLoading: true,
    error: null,
  };
  it("renders correctly on load", () => {
    useActionWebInvoke.mockReturnValue(mockResponse);
    const { asFragment } = render(<PlacementList {...fakeProps} />);
    expect(asFragment()).toMatchSnapshot();
  });
  it("has a <ProgressCircle> when loading", () => {
    useActionWebInvoke.mockReturnValue(mockResponse);
    render(<PlacementList {...fakeProps} />);
    // screen.debug();
    expect(screen.getByLabelText("Getting Offer Placement")).toBeDefined();
  });
});

describe("<PlacementList> on load", () => {
  const mockResponse = {
    data: mock.data.offerPlacements,
    isLoading: false,
    error: null,
  };
  it("renders correctly on data", async () => {
    useActionWebInvoke.mockReturnValue(mockResponse);
    render(
      <Provider theme={defaultTheme} colorScheme={`light`}>
        <PlacementList {...fakeProps} />
      </Provider>
    );
    expect(screen.getByText("Select Offer Placement")).toBeInTheDocument();
    userEvent.click(screen.getByLabelText("Select Offer Placement"));
  });
});

describe("<PlacementList> on error", () => {
  const mockResponse = {
    data: null,
    isLoading: false,
    error: { message: "fake-error-message" },
  };
  it("renders correctly on error", () => {
    useActionWebInvoke.mockReturnValue(mockResponse);
    const { asFragment } = render(<PlacementList {...fakeProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("show error message", async () => {
    useActionWebInvoke.mockReturnValue(mockResponse);
    render(<PlacementList {...fakeProps} />);
    expect(screen.getAllByText("fake-error-message")).toBeDefined();
  });
});

describe("<PlacementList> on data", () => {
  const mockResponse = {
    data: null,
    isLoading: false,
    error: null,
  };
  it("renders correctly on no data", () => {
    useActionWebInvoke.mockReturnValue(mockResponse);
    const { asFragment } = render(<PlacementList {...fakeProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("show no data message", async () => {
    useActionWebInvoke.mockReturnValue(mockResponse);
    render(<PlacementList {...fakeProps} />);
    expect(screen.getAllByText("You have no placements !")).toBeDefined();
  });
});
