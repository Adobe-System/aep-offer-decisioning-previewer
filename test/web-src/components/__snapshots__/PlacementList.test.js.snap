// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlacementList> calls custom hook call custom hooks with global parameters correctly 1`] = `
<DocumentFragment>
  <div
    aria-label="Getting Offer Placement"
    aria-valuemax="100"
    aria-valuemin="0"
    class="_spectrum-CircleLoader_40b43 _spectrum-CircleLoader--indeterminate_40b43"
    id="placement-list-progress-circle"
    role="progressbar"
  >
    <div
      class="_spectrum-CircleLoader-track_40b43"
    />
    <div
      class="_spectrum-CircleLoader-fills_40b43"
    >
      <div
        class="_spectrum-CircleLoader-fillMask1_40b43"
      >
        <div
          class="_spectrum-CircleLoader-fillSubMask1_40b43"
        >
          <div
            class="_spectrum-CircleLoader-fill_40b43"
          />
        </div>
      </div>
      <div
        class="_spectrum-CircleLoader-fillMask2_40b43"
      >
        <div
          class="_spectrum-CircleLoader-fillSubMask2_40b43"
        >
          <div
            class="_spectrum-CircleLoader-fill_40b43"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<PlacementList> calls custom hook call custom hooks with global parameters correctly 2`] = `
<DocumentFragment>
  <div
    aria-label="Getting Offer Placement"
    aria-valuemax="100"
    aria-valuemin="0"
    class="_spectrum-CircleLoader_40b43 _spectrum-CircleLoader--indeterminate_40b43"
    id="placement-list-progress-circle"
    role="progressbar"
  >
    <div
      class="_spectrum-CircleLoader-track_40b43"
    />
    <div
      class="_spectrum-CircleLoader-fills_40b43"
    >
      <div
        class="_spectrum-CircleLoader-fillMask1_40b43"
      >
        <div
          class="_spectrum-CircleLoader-fillSubMask1_40b43"
        >
          <div
            class="_spectrum-CircleLoader-fill_40b43"
          />
        </div>
      </div>
      <div
        class="_spectrum-CircleLoader-fillMask2_40b43"
      >
        <div
          class="_spectrum-CircleLoader-fillSubMask2_40b43"
        >
          <div
            class="_spectrum-CircleLoader-fill_40b43"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<PlacementList> on data renders correctly on no data 1`] = `
<DocumentFragment>
  <span>
    You have no placements !
  </span>
</DocumentFragment>
`;

exports[`<PlacementList> on error renders correctly on error 1`] = `
<DocumentFragment>
  <span>
    fake-error-message
  </span>
</DocumentFragment>
`;

exports[`<PlacementList> on load renders correctly on load 1`] = `
<DocumentFragment>
  <div
    aria-label="Getting Offer Placement"
    aria-valuemax="100"
    aria-valuemin="0"
    class="_spectrum-CircleLoader_40b43 _spectrum-CircleLoader--indeterminate_40b43"
    id="placement-list-progress-circle"
    role="progressbar"
  >
    <div
      class="_spectrum-CircleLoader-track_40b43"
    />
    <div
      class="_spectrum-CircleLoader-fills_40b43"
    >
      <div
        class="_spectrum-CircleLoader-fillMask1_40b43"
      >
        <div
          class="_spectrum-CircleLoader-fillSubMask1_40b43"
        >
          <div
            class="_spectrum-CircleLoader-fill_40b43"
          />
        </div>
      </div>
      <div
        class="_spectrum-CircleLoader-fillMask2_40b43"
      >
        <div
          class="_spectrum-CircleLoader-fillSubMask2_40b43"
        >
          <div
            class="_spectrum-CircleLoader-fill_40b43"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
